File: prototype-python/src/generator.py
import random

class BingoGenerator:
    def generate_cards(self, songs, num_cards, jackpot=False):
        """
        Generates 'num_cards' unique bingo grids.
        songs: list of {'title': str, 'artist': str}
        Returns: list of 5x5 matrices (lists of lists of dicts).
        """
        # Deduplicate songs based on title and artist (case-insensitive)
        unique_songs = []
        seen = set()
        for s in songs:
            title_clean = s['title'].strip().lower()
            artist_clean = s['artist'].strip().lower() if s['artist'] else ""
            identifier = (title_clean, artist_clean)
            if identifier not in seen:
                seen.add(identifier)
                unique_songs.append(s)

        min_required = 25 if jackpot else 24
        if len(unique_songs) < min_required:
            raise ValueError(f"Not enough unique songs! Need at least {min_required}, but only got {len(unique_songs)}.")

        cards = []
        for _ in range(num_cards):
            # 1. Select unique songs
            selection = random.sample(unique_songs, min_required)
            
            # 2. Shuffle the selection for grid placement
            random.shuffle(selection)
            
            # 3. Insert FREE SPACE at index 12 (Center of 5x5) only if jackpot is disabled
            if not jackpot:
                selection.insert(12, {'title': "FREE SPACE", 'artist': ""})
            
            # 4. Reshape into 5x5 grid
            grid = []
            for i in range(0, 25, 5):
                row = selection[i : i+5]
                grid.append(row)
            
            cards.append(grid)
            
        return cards

File: mobile-app/src/utils/generator.ts
// Bingo Card Generator Logic

export interface Song {
  title: string;
  artist: string;
}

export interface BingoCard {
  id: number;
  grid: Song[][]; // 5x5 grid
}

export const generateBingoCards = (
  songs: Song[],
  numCards: number,
  jackpot: boolean = false
): BingoCard[] => {
  const minRequired = jackpot ? 25 : 24;
  if (songs.length < minRequired) {
    throw new Error(
      `Not enough unique songs! Need at least ${minRequired}, but got ${songs.length}.`
    );
  }

  const cards: BingoCard[] = [];

  for (let i = 0; i < numCards; i++) {
    // 1. Select needed unique songs (shuffle and slice)
    // Create a copy to shuffle
    const shuffledPool = [...songs].sort(() => 0.5 - Math.random());
    const selection = shuffledPool.slice(0, minRequired);

    // 2. Shuffle the selection for grid placement
    const cardSongs = selection.sort(() => 0.5 - Math.random());

    // 3. Insert FREE SPACE at index 12 only if jackpot is disabled
    if (!jackpot) {
      const freeSpace: Song = { title: "FREE SPACE", artist: "" };
      cardSongs.splice(12, 0, freeSpace);
    }

    // 4. Reshape into 5x5 grid
    const grid: Song[][] = [];
    for (let row = 0; row < 5; row++) {
      grid.push(cardSongs.slice(row * 5, row * 5 + 5));
    }

    cards.push({ id: i + 1, grid });
  }

  return cards;
};
